force-app/main/default/classes/DMLOperations.cls

```diff
@@ -161,34 +197,43 @@ public with sharing class DMLOperations {
 	public static Account upsertAccount(String accountName) {
 		// Query for the Account
 		List<Account> accounts = [SELECT Id FROM Account WHERE Name = :accountName];
+        Account account = new Account();
 
 		// if the Account exists, update the description
-
 		// otherwise, create a new Account and set the description
-
-		// Upsert the Account
-		return null;
+        if (!accounts.isEmpty()) {
+            account = accounts[0];
+            account.description = 'Updated Account';
+        } else {
+            account.Name = accountName;
+            account.Description = 'New Account';
+        }
+
+        // Upsert the Account
+        upsert account;
+        return account;
 	}
 
     /**
      * Question 10
 	 * Manage the association between Contacts and Accounts
 	 * Ensure that for each Contact in the provided list, there is a corresponding Account Name that matches the Contact's last name.
-	 * If such an Account does not exist, the method creates it; if it does, the method ensures the Contact is associated with it. (Hint: use previous upsertAccount method - this may cause bad practices but for learning purposes it's okay.)
+	 * If such an Account does not exist, the method creates it; if it does, the method ensures the Contact is associated with it.
+     *      (Hint: use previous upsertAccount method - this may cause bad practices but for learning purposes it's okay.)
      * Example Contact.LastName = 'Doe' Link to Account.Name = 'Doe' 
      * Example Contact.LastName = 'Jane' Link to Account.Name = 'Jane'
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
 		// Loop through each Contact
-
 		// Extract the Account Name from Contact's LastName
-
 		// Get the account based on the name or create one if it doesn't exist
-
 		// Link the Contact to its Account using the Account Id
-
 		// Upsert the Contacts
+        for (Contact contact : contactsToUpsert) {
+            contact.accountid = upsertAccount(contact.LastName).Id;
```

Comment:
We will let this slide because the directions say this is ok, but for practice, see if you can do it without DML inside the loop. Maybe refactor the method for question 9.

Reply:
I'll make a note on this to do later. Need to keep moving/catching up on subsequent homeworks.